artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java

```diff
@@ -278,21 +293,26 @@ public boolean isPaging() {
       lock.readLock().lock();
 
       try {
-         if (addressFullMessagePolicy == AddressFullMessagePolicy.BLOCK) {
-            return false;
-         }
-         if (addressFullMessagePolicy == AddressFullMessagePolicy.FAIL) {
-            return isFull();
-         }
-         if (addressFullMessagePolicy == AddressFullMessagePolicy.DROP) {
-            return isFull();
-         }
-         return paging;
+         return isPagingDirtyRead();
       } finally {
          lock.readLock().unlock();
       }
    }
 
+   @Override
+   public boolean isPagingDirtyRead() {
+      if (addressFullMessagePolicy == AddressFullMessagePolicy.BLOCK) {
```

Comment:
we can read it just once and save it in a local variable, avoiding 3 volatile loads: same can be done on the original version too

Reply:
nice idea!

Reply:
addressFullMessagePolicy would be changed if address setting is reapplied. so we need to load the value.

Reply:
Yes, but we can just volatile load once before checking its value 3 times, on each call of  isPagingDirtyRead

Reply:
@wy96f what @franz1981 is trying to say, is we can do the volatile read just once, by adding one line e.g.


         AddressFullMessagePolicy addressFullMessagePolicy = this.addressFullMessagePolicy;
         if (addressFullMessagePolicy == AddressFullMessagePolicy.BLOCK) {
            return false;
         }
         if (addressFullMessagePolicy == AddressFullMessagePolicy.FAIL) {
            return isFull();
         }
         if (addressFullMessagePolicy == AddressFullMessagePolicy.DROP) {
            return isFull();
         }
         return paging;

Reply:
> Yes, but we can just volatile load once before checking its value 3 times, on each call of isPagingDirtyRead

get it. nice catch :+1: